File: bubble.py
import random
import sys
import time


def bubble(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr


def main(threads, n):
    arr = [random.randint(0, n - 1) for _ in range(n)]
    if threads == 1:
        bubble(arr[:])
    else:
        from concurrent.futures import ThreadPoolExecutor

        with ThreadPoolExecutor(max_workers=threads) as executor:
            for _ in range(threads):
                executor.submit(bubble, arr[:])


if __name__ == '__main__':
    label = sys.argv[1]
    threads = int(sys.argv[2])
    n = int(sys.argv[3])
    sum = 0
    for _ in range(3):
        start = time.time()
        main(threads, n)
        end = time.time()
        sum += end - start
    avg = sum / 3
    print(f'{avg:.04f} {label}')

File: bubble.js
const bubble = (arr) => {
  const n = arr.length;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - i - 1; j++) {
      const a = arr[j];
      const b = arr[j + 1];
      if (a > b) {
        arr[j] = b;
        arr[j + 1] = a;
      }
    }
  }
  return arr;
}

const getRandomInt = max => {
  return Math.floor(Math.random() * max);
}

const label = process.argv[2];
const n = process.argv[3];
let sum = 0;
for (let i = 0; i < 3; i++) {
  const arr = Array.from({length: n}, _ => getRandomInt(n));
  const start = Number(process.hrtime.bigint() / 100000n) / 10000;
  bubble(arr);
  const end = Number(process.hrtime.bigint() / 100000n) / 10000;
  sum += end - start;
}
console.log(parseInt(sum / 3 * 10000) / 10000, label);
